Creation Modal for Course Instances Page

miguelaenlle opened the pull request:
Closes #10964. 

- Developed a Create Course Instance modal that lets users type in a CIID, Long Name, and (optionally) a Start Access Date and End Access Date:

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/ae586eb5-85eb-47bd-b322-cba3d0229c20">

- Implemented testing for the Modal

reteps commented:
Question for @nwalters512 : At what point do we want to get the Preact PR merged in so we can add some frontend validation e.g. react-hook-form alongside our Zod schemas? Some of these forms (maybe not this one as much) will benefit from frontend form validation for a better UX.

nwalters512 commented:
@reteps We already merged the PR that lays the groundwork for using Preact on the client: #10800. I agree that some validation would be nice! Even in this PR, we can validate that the end date is after the start date, and maybe validate that either both dates are set, or that neither are. I'm ambivalent about using React in this case, as it's just as easy to do the validation with plain JavaScript, though I wouldn't mind using Preact here. @miguelaenlle any thoughts here?

miguelaenlle commented:
> @reteps We already merged the PR that lays the groundwork for using Preact on the client: #10800. I agree that some validation would be nice! Even in this PR, we can validate that the end date is after the start date, and maybe validate that either both dates are set, or that neither are. I'm ambivalent about using React in this case, as it's just as easy to do the validation with plain JavaScript, though I wouldn't mind using Preact here. @miguelaenlle any thoughts here?

I'd prefer to implement the modal and validation functionality in plain JS for this PR, to keep the validation implementation simpler and to keep the modal on this page consistent with the modals throughout PrairieLearn.

However, let me know if there's an easy/more beneficial way to implement this with Preact.

nwalters512 commented:
I think that's the right move for now! Our support for Preact is still in its early stages, and I don't want to get hung up on trying to figure out novel issues as part of this PR.

miguelaenlle commented on apps/prairielearn/src/lib/editors.ts:
```diff
@@ -489,47 +491,66 @@ export abstract class Editor {
-          const foundNumber = parseInt(found[1]);
-          if (foundNumber >= number) {
-            number = foundNumber + 1;
+          const foundNumber = parseInt(found[1]) || 1;
```
found[1] on a boolean isn't valid syntax...

miguelaenlle commented on apps/prairielearn/src/lib/editors.ts:
```diff
@@ -489,47 +491,66 @@ export abstract class Editor {
-    };
+
+    if (number === 1 && shortName !== 'New' && longName !== 'New') {
+      // If there are no existing copies, and the shortName/longName aren't the default ones, we don't need to add a number
```
Remove 'we'

miguelaenlle commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.ts:
```diff
@@ -57,13 +59,24 @@ router.post(
         await editor.executeWithServerJob(serverJob);
-      } catch {
+      } catch (e) {
+        console.error(e);
```
Remove this

miguelaenlle commented on apps/prairielearn/src/tests/courseEditor.test.ts:
```diff
@@ -29,11 +29,17 @@ const courseDir = courseLiveDir;
-const courseInstancesUrl = `${courseInstanceUrl}/course_admin/instances`;
+
+const newCourseInstanceUrl = baseUrl + '/course_instance/2/instructor';
+const settingsUrl = `${newCourseInstanceUrl}/instance_admin/settings`;
```
Call this "newCourseInstanceSettingsUrl"

miguelaenlle commented on apps/prairielearn/src/tests/editorGetNamesForAdd.test.ts:
```diff
@@ -0,0 +1,121 @@
+});
+
+describe('Editor getNamesForAdd Tests', () => {
+  step('No specified short_name should default to New', () => {});
```
Implement this test

miguelaenlle commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +232,68 @@ function PopoverEndDate() {
+          type="datetime-local"
+          id="start_access_date"
+          name="start_access_date"
+          value=""
```
Hi @mwest1066 . I added this functionality:

<img width="876" alt="image" src="https://github.com/user-attachments/assets/bad1cd60-3ca0-49d8-a671-33954435fd64" />

mwest1066 replied:
Looks great!

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+      </div>
+      <div class="form-group">
+        <label for="start_access_date">Start Access Date</label>
+        <input
```
Adding `step="1"` should help ensure that the input always shows with second resolution.

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+          This is the full name of the course instance, such as "Fall 2019" or "Spring 2020".
+        </small>
+      </div>
+      <div class="form-group">
```
For consideration: immediately forcing someone to pick when a course instance should be accessible to students might make them bounce if they just want to experiment with the platform. Maybe we should have a "Make course instance available to students" checkbox that will disable and hide the date inputs? cc @mfsilva22 I'd like your input here.

miguelaenlle replied:
I'll wait for @mfsilva22 's input prior to making this change.

mfsilva22 replied:
I agree that we should not force instructors to enter the dates at first. A checkbox is a good idea. I would be more inclined to not request for dates as default. We can add some help text saying the dates can be edited later.

miguelaenlle replied:
I added the "Make course instance available to students" checkbox:

<img width="1079" alt="image" src="https://github.com/user-attachments/assets/8767a0e8-746b-4f7a-8b61-fd78d69c6042" />

<img width="1078" alt="image" src="https://github.com/user-attachments/assets/c9e2d67b-8738-4d8f-a49e-7754681c1227" />

nwalters512 replied:
Excellent 🙌

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+          name="short_name"
+          required
+          pattern="[\\-A-Za-z0-9_\\/]+"
+          title="Please enter a valid format: only letters, numbers, hyphens, underscores, backslashes, and forward slashes."
```
`title` is generally not considered very accessible: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/title#accessibility_concerns. I think we can just drop this; the label below the form should be sufficient.

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+          pattern="[\\-A-Za-z0-9_\\/]+"
+          title="Please enter a valid format: only letters, numbers, hyphens, underscores, backslashes, and forward slashes."
+        />
+        <small class="form-text text-muted"
+          >The recommended format is <code>Fa19</code> or <code>Fall2019</code>. Use only letters,
```
Nit:

```suggestion
        <small class="form-text text-muted">
          The recommended format is <code>Fa19</code> or <code>Fall2019</code>. Use only letters,
```

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+  initialEndDateFormatted,
+}: {
+  csrfToken: string;
+  initialStartDateFormatted?: string;
+  initialEndDateFormatted?: string;
```
These should always be present; can we make them non-optional (drop the `?`)?

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+          pattern="[\\-A-Za-z0-9_\\/]+"
+          title="Please enter a valid format: only letters, numbers, hyphens, underscores, backslashes, and forward slashes."
+        />
+        <small class="form-text text-muted"
```
You should use `aria-describedby` to link these bits of explanatory text with the input; this will help out folks who are using screen readers. There's an example of this on the email input in this example in the Bootstrap docs: https://getbootstrap.com/docs/5.3/forms/overview/#overview

miguelaenlle replied:
I added this to the inputs and the `small` text descriptions. Here's the updated short_name input, for example:

![image](https://github.com/user-attachments/assets/fae1d6a5-adf6-4072-bb55-31b1c6edc784)

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+}) {
+  return Modal({
+    id: 'createCourseInstanceModal',
+    title: 'Create Course Instance',
```
Conventionally we use sentence case here:

```suggestion
    title: 'Create course instance',
```

nwalters512 commented on apps/prairielearn/src/tests/editorGetNamesForAdd.test.ts:
```diff
@@ -0,0 +1,125 @@
+});
+
+describe('Editor getNamesForAdd Tests', () => {
+  step('No specified short_name and long_name should default them to New_1', () => {
```
Our use of `mocha-steps` in the integration tests is only because we use each step to build up state that the following tests rely on. For unit tests like this that don't share state, it's generally better to just use `it(...)`. See here for an example: https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/src/lib/json.test.ts

miguelaenlle replied:
Got it! I made this change in `editors.test.ts` (the new file for these tests)

nwalters512 commented on apps/prairielearn/src/lib/editors.ts:
```diff
@@ -889,16 +925,53 @@ export class CourseInstanceAddEditor extends Editor {
 
+    let allowAccess: { startDate?: string; endDate?: string } | undefined = undefined;
+
+    let startDate: Temporal.ZonedDateTime | undefined;
+    let endDate: Temporal.ZonedDateTime | undefined;
+
+    if (this.start_access_date) {
+      startDate = Temporal.PlainDateTime.from(this.start_access_date).toZonedDateTime(
+        this.course.display_timezone,
+      );
+    }
+
+    if (this.end_access_date) {
+      endDate = Temporal.PlainDateTime.from(this.end_access_date).toZonedDateTime(
+        this.course.display_timezone,
+      );
+    }
+
+    if (startDate && endDate && startDate.epochMilliseconds > endDate.epochMilliseconds) {
+      throw new HttpStatusError(400, 'Start date must be before end date');
+    }
```
I'd prefer to push this parsing and validation higher up in the stack. Once we start the edit/sync process, we really don't want it to fail. That is: I think `start_access_date`/`end_access_date` as received in the `params` here should themselves be `Date` or maybe some type of `Temporal` instance.

miguelaenlle replied:
I modified `CourseInstanceAddEditor` to receive `start_access_date` and `end_access_date` as `Temporal.ZonedDateTime` objects, and moved their validation to the constructor:

![image](https://github.com/user-attachments/assets/a7edd2fb-91ff-4f9b-b5eb-04b14ee7b7a2)

I also moved the date parsing to `instructorCourseAdminInstances.ts`, within the POST request function. There is also validation here to ensure that end_access_date comes after start_access_date.

![image](https://github.com/user-attachments/assets/d3ee75ea-3098-40a5-8829-872232fc6a9d)

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +253,78 @@ function PopoverEndDate() {
+          name="start_access_date"
+          value="${initialStartDateFormatted}"
+          max="${initialEndDateFormatted}"
+        />
```
In other contexts, we like to explicitly show the time zone that's being used, e.g.:

<img width="390" alt="Screenshot 2024-12-13 at 14 16 08" src="https://github.com/user-attachments/assets/45900090-e0f3-426b-92b9-e394637f6da7" />

That's implemented here:

https://github.com/PrairieLearn/PrairieLearn/blob/81a771b6e6d9b7cce6aa80851efa1c4d9da4ce59/apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx#L595

Might be nice to do the same here as well!

miguelaenlle replied:
I added this:

![image](https://github.com/user-attachments/assets/92e998b9-a3a1-49c4-9d75-c9b935dee17c)

nwalters512 commented on apps/prairielearn/src/tests/instructorCourseAdminInstances.test.ts:
```diff
@@ -0,0 +1,255 @@
+
+    assert.equal(courseInstanceInfo.longName, 'Fall 2019');
+    assert.equal(courseInstanceInfo.allowAccess.length, 1);
+    assert.equal(courseInstanceInfo.allowAccess[0].startDate, '2021-01-01 00:00:00');
+    assert.equal(courseInstanceInfo.allowAccess[0].endDate, '2021-01-02 00:00:00');
```
cc @mwest1066 the current implementation is canonicalizing to the datetime format without `T` in the middle. Do you have any preference either way? I sort of lean towards including the `T` because it's familiar from ISO strings, but maybe the average instructor would prefer not to have them.

mwest1066 replied:
Great question. If we can read the date times without the T then I think this is friendlier for most people. I think most instructors have never heard of ISO 8601 😂

nwalters512 replied:
SGTM! The `input_date` sproc does explicitly handle this case:

https://github.com/PrairieLearn/PrairieLearn/blob/b8cb11c28489bdad4b0e7d6b18ed6939442d4e93/apps/prairielearn/src/sprocs/input_date.sql#L16-L18

nwalters512 reviewed (commented):
Here's a bunch of comments! As with the last PR, I'm perhaps being extra picky because we really want a smooth onboarding experience, and because the patterns applied here will be used across several upcoming pages.

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorCourseAdminInstancesClient.ts:
```diff
@@ -0,0 +1,15 @@
+  });
+  endAccessDate.addEventListener('change', () => {
+    startAccessDate.max = endAccessDate.value;
+  });
```
Hi @nwalters512 . I added this change -- on an attempted submission, if startAccessDate >= endAccessDate, custom validity on the endAccessDateInput is set to 'End access date must be after start access date':

![image](https://github.com/user-attachments/assets/b76fccc7-c387-4672-b731-bcfffc65ca7f)

nwalters512 replied:
Looks great!

miguelaenlle commented on apps/prairielearn/src/tests/editorGetNamesForAdd.test.ts:
```diff
@@ -0,0 +1,125 @@
+
+describe('Editor getNamesForAdd Tests', () => {
+  step('No specified short_name and long_name should default them to New_1', () => {
+    const names = editor.getNamesForAdd(['Fa18', 'Fa19'], ['Fall 2018', 'Fall 2019']);
```
Agreed that this should be a regular function. I moved `getNamesForAdd` out of the `Editor` class and moved the tests to `editors.test.ts`.

mfsilva22 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +254,97 @@ function PopoverEndDate() {
+    formMethod: 'POST',
+    body: html`
+      <div class="form-group">
+        <label for="short_name">CIID (Course Instance Identifier)</label>
```
Why not `Course Instance Identifier (CIID)`?

mfsilva22 commented:
Few things to consider:
- `Long Name` input field first. `CIID` next.
- It would be nice to provide a short description of these two items (why are they inputing similar information twice?). One is a string that will be concatenated with the course name, to become the name of the course instance (eg. MATH 101, Fall 2025). This is what is visible to users. The other one is just the unique name for internal use. I do not know how to easily explain this, but if I try to pretend I am using PL for the first time, I would be confused about the "repeated" request for information.

miguelaenlle commented:
> Few things to consider:
> 
> * `Long Name` input field first. `CIID` next.
> * It would be nice to provide a short description of these two items (why are they inputing similar information twice?). One is a string that will be concatenated with the course name, to become the name of the course instance (eg. MATH 101, Fall 2025). This is what is visible to users. The other one is just the unique name for internal use. I do not know how to easily explain this, but if I try to pretend I am using PL for the first time, I would be confused about the "repeated" request for information.

Hi @mfsilva22 . I added swapped the order of the course instance name input fields, and updated their descriptions. Let me know if you have any suggestions for modification to them:

<img width="986" alt="image" src="https://github.com/user-attachments/assets/068dc9d2-0edc-42fa-a1ff-91240e0a72eb" />

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -16,11 +21,33 @@ export function InstructorCourseAdminInstances({
   resLocals: Record<string, any>;
   courseInstances: CourseInstanceAuthzRow[];
 }) {
+  const startDate = Temporal.Now.zonedDateTimeISO(resLocals.course.timeZone).with({
+    hour: 0,
+    minute: 1,
+    second: 0,
+  });
+  const initialStartDateFormatted = formatDateYMDHM(
+    new Date(startDate.epochMilliseconds),
+    resLocals.course.time_zone,
+  );
+
+  const endDate = startDate.add({ months: 4 }).with({
+    hour: 23,
+    minute: 59,
+    second: 0,
+  });
+
+  const initialEndDateFormatted = formatDateYMDHM(
+    new Date(endDate.epochMilliseconds),
+    resLocals.course.time_zone,
+  );
```
Nits (mainly name consistency to avoid confusion):

```suggestion
  const initialStartDate = Temporal.Now.zonedDateTimeISO(resLocals.course.timeZone).with({
    hour: 0,
    minute: 1,
    second: 0,
  });
  const initialStartDateFormatted = formatDateYMDHM(
    new Date(initialStartDate.epochMilliseconds),
    resLocals.course.time_zone,
  );

  const initialEndDate = initialStartDate.add({ months: 4 }).with({
    hour: 23,
    minute: 59,
    second: 0,
  });
  const initialEndDateFormatted = formatDateYMDHM(
    new Date(initialEndDate.epochMilliseconds),
    resLocals.course.time_zone,
  );
```

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +255,118 @@ function PopoverEndDate() {
+          aria-describedby="short_name_help"
+        />
+        <small id="short_name_help" class="form-text text-muted">
+          ID for internal use, of format
```
Nit:

```suggestion
          ID for internal use, e.g.
```

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +255,118 @@ function PopoverEndDate() {
+          name="access_dates_enabled"
+          aria-describedby="access_dates_enabled_help"
+        />
+        <label class="form-check-label" for="access_dates_enabled"
+          >Make course instance available to students</label
+        >
```
```suggestion
        <label class="form-check-label" for="access_dates_enabled">
          Make course instance available to students
        </label>
```

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +255,118 @@ function PopoverEndDate() {
+      </div>
+      <div id="accessDates" hidden="true">
+        <div class="form-group">
+          <label for="start_access_date">Start Access Date</label>
```
Nit, this is how we usually write this (you can change the end date too, and probably the input IDs and names as well):

```suggestion
          <label for="start_access_date">Access Start Date</label>
```

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +255,118 @@ function PopoverEndDate() {
+        </small>
+      </div>
+      <div class="form-group">
+        <label for="short_name">Course Instance Identifier (CIID)</label>
```
@mwest1066 we don't really use the term "CIID" elsewhere the way we do e.g. "QID" (though we do use it on the course instance settings page). We call this "short name" in the database. Any thoughts on what we should do here?

mwest1066 replied:
I think we should avoid the term CIID because we don't use it elsewhere. "Short name" or "abbreviation"? Maybe we should change the explanation to something like:
> A short name (less than 8 letters) such as "Fa25" or "W25b". This is used in menus and headers where a short description is required.

nwalters512 replied:
I think "short name" might be best.

We don't enforce any length restrictions, so I'm inclined not to mention anything about length here. We have lots of people that are using longer names in practice. Some people are integrating section/instructor names in here, which IMO is totally valid.

jonatanschroeder replied:
Just raising #9908. Not necessarily saying this needs to be addressed here, but something to keep in mind. The issue raises the option of removing that requirement, but keeps open the option of reviewing them to ensure they are reasonable (e.g., no `//`, no paths including `.git`, possibly avoid non-ASCII characters), and in such cases this may be a valid option to consider here, since the CIID is also a directory name.

mwest1066 replied:
Good point, I think restricting characters/format would be a good idea.

nwalters512 replied:
@mwest1066 this PR does currently restrict the characters, see the `pattern` attribute on the input. I think Jonatan was actually suggesting _loosening_ the restriction. IMO, we should save that for when we can more comprehensively address #9908 if we choose to.

nwalters512 replied:
On that note - @miguelaenlle IIRC a previous iteration called out that this can only use letters/numbers/dashes/underscores without spaces. I think that was nice, as otherwise there's no way for users to discover that except through trial and error.

mfsilva22 replied:
<img width="994" alt="image" src="https://github.com/user-attachments/assets/2a68a853-c28a-4d1d-8602-68c8f5268c5c" />

We do use CIID. I was going to suggest changing the name as well, but stopped myself once I noticed that. maybe update this other page too?

mwest1066 replied:
I think we should change that page too. I think `QID` is useful terminology because we actually use them to refer to questions, but `AID`/`QIID` shouldn't be used (we should use "short name" instead). I think "Name" and "Short name" should be our standard terminology. (that is, "Name" rather than "Long name")

mwest1066 replied:
I guess I think that "title" sounds a bit strange for a course instance, while it's reasonable for questions and assessments. I don't have a great argument about why this is, however 😅. And I also agree for consistency "title" for everything is probably the best solution.

And I agree this is way out of scope for this PR!

nwalters512 replied:
Yeah, agreed that "title" for a course instance does sounds weird. Maybe scratch that and we just don't worry about that particular inconsistency.

Think we should change to just "name" in this PR? Or save this for another one where we can make this all consistent?

miguelaenlle replied:
> On that note - @miguelaenlle IIRC a previous iteration called out that this can only use letters/numbers/dashes/underscores without spaces. I think that was nice, as otherwise there's no way for users to discover that except through trial and error.

I updated the Short name field description to also have instructions on the format to use:

<img width="796" alt="image" src="https://github.com/user-attachments/assets/a021d355-787a-400c-9dde-3063ab22ec30" />

nwalters512 replied:
Talked with @miguelaenlle about this offline, we're going to save any larger naming changes for another PR where we can make those changes holistically.

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +255,118 @@ function PopoverEndDate() {
+    formMethod: 'POST',
+    body: html`
+      <div class="form-group">
+        <label for="long_name">Long Name</label>
```
Nit, we typically use sentence case for input labels. Can you change the others as well?

```suggestion
        <label for="long_name">Long name</label>
```

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.ts:
```diff
@@ -81,7 +121,7 @@ router.post(
           '/course_instance/' +
           courseInstanceId +
-          '/instructor/instance_admin/settings',
+          '/instructor/instance_admin/assessments',
```
Consider adding a flash message here to let users know that everything went ok. You can see an example of that here:

https://github.com/PrairieLearn/PrairieLearn/blob/b8cb11c28489bdad4b0e7d6b18ed6939442d4e93/apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts#L171

Consider text like "Course instance created successfully."

miguelaenlle replied:
I added this:

<img width="1700" alt="image" src="https://github.com/user-attachments/assets/05ed03fe-7558-4695-992f-b5420c99cc1d" />

nwalters512 replied:
Nice! Can you add a period at the end? I know the example I linked to doesn't include one, but I think that's an oversight.

miguelaenlle replied:
Yes! I added this change:

<img width="870" alt="image" src="https://github.com/user-attachments/assets/51cee244-cf36-4633-ac17-c5bb6d550a30" />

nwalters512 replied:
Perfect!

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -16,11 +21,33 @@ export function InstructorCourseAdminInstances({
   resLocals: Record<string, any>;
   courseInstances: CourseInstanceAuthzRow[];
 }) {
+  const startDate = Temporal.Now.zonedDateTimeISO(resLocals.course.timeZone).with({
+    hour: 0,
+    minute: 1,
+    second: 0,
+  });
+  const initialStartDateFormatted = formatDateYMDHM(
+    new Date(startDate.epochMilliseconds),
+    resLocals.course.time_zone,
+  );
+
+  const endDate = startDate.add({ months: 4 }).with({
+    hour: 23,
+    minute: 59,
+    second: 0,
+  });
```
We use these defaults in a few other places. @mwest1066 any preference here?

```suggestion
  const startDate = Temporal.Now.zonedDateTimeISO(resLocals.course.timeZone).with({
    hour: 0,
    minute: 0,
    second: 1,
  });
  const initialStartDateFormatted = formatDateYMDHM(
    new Date(startDate.epochMilliseconds),
    resLocals.course.time_zone,
  );

  const endDate = startDate.add({ months: 4 }).with({
    hour: 23,
    minute: 59,
    second: 59,
  });
```

mwest1066 replied:
We've previously used 00:00:01 and 23:59:59 so that it's basically midnight from either side. However, I think we rarely (if ever) want seconds level of precision in our times, and it adds a lot of visual noise throughout the interface. PT has consistently used 00:01 and 23:59 with date inputs set to use minutes as the step size, and we haven't heard any complaints. I'm wondering whether we can do the same in PL? The only problem I see is one of backwards compatibility with pre-existing dates in access rules.

nwalters512 replied:
Hmm, interesting! Certainly we could do something where we only display seconds for editing if they're non-zero. This may not even apply to this PR, where we're always creating things from scratch, so if you want to move in the direction of minute precision by disabling seconds here, I'd be OK with that.

nwalters512 reviewed (commented):
This is looking great! I played with it locally and everything is pretty much working as expected. A few more things to talk through before we merge this.

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorCourseAdminInstancesClient.ts:
```diff
@@ -0,0 +1,46 @@
+
+  if (!startAccessDateInput || !endAccessDateInput || !createButton) return;
+
+  createButton.onclick = () => {
```
I replaced the `onclick` and `onchange` with `addEventListener` in this file:

![image](https://github.com/user-attachments/assets/34ebacb0-a298-419e-acf2-dc0f10cbad7f)

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:
```diff
@@ -225,3 +254,115 @@ function PopoverEndDate() {
+          Make course instance available to students
+        </label>
+        <small id="access_dates_enabled_help" class="form-text text-muted">
+          This can be enabled later.
```
Nit: do you think this might look better on a separate line?

<img width="371" alt="Screenshot 2024-12-16 at 14 29 13" src="https://github.com/user-attachments/assets/5fc74b87-0d39-4c8a-8f8d-13d2548d14f7" />

If we do this, we can probably also drop the conditional show/hide and just always show this, as indeed student access can be changed (removed entirely) later.

miguelaenlle replied:
I agree, for consistency with the other input fields. 

Here's what it looks like now:

<img width="407" alt="image" src="https://github.com/user-attachments/assets/15d7117b-ee1a-432c-973e-6f364b4bc0e7" />

The "this can be enabled later." text is also always shown:

<img width="497" alt="image" src="https://github.com/user-attachments/assets/7f88011f-cded-48fa-9171-8cd7d504abbf" />

nwalters512 replied:
Looks great 💯
